### DISCUSSION TITLE
Issue with calculating initial currents in a 2P configuration

### DISCUSSION THREAD

**Discussion by MarcoZangrandi** [pybamm-team/PyBaMM#4078] [Q&A]
Hi everyone,
I'm trying to build a 2P configuration with two Chen2020-type cells to observe current dynamics. 

The starting hypothesis is that the two cells have different State of Charge (SOC): one is fully charged at 100% and the other at 80%. I'm using an SPMe model to which I provide a current at each step. From the model, I need to extract the Open-Circuit Voltage (OCV) and internal resistance. The two cells are thus seen as a voltage generator with an internal resistance in series.

These values, via an algorithm, will update the currents passing through the two cells. The current calculation algorithm is based on current formulas present in the article "[Current distribution of parallel-connected cells in dependence of cell resistance, capacity and number of parallel cells.](https://www.sciencedirect.com/science/article/abs/pii/S0378775318311698)"

The problem I encountered is the following. Wanting to simulate a real-life scenario, initially, as I mentioned, I have two cells charged differently (in the future, this will be generalized for cells with different chemistries). Consequently, at the initial instant, they will have two different internal resistances. With two different internal resistances, the current I supply to the parallel connection will be redistributed proportionally to ensure that "V<sub>ocv_cell</sub> – R<sub>internal_cell</sub>*I<sub>cell</sub>" yields the "Terminal voltage". Secondly, the constraint to always respect is that the two "Terminal voltages" must always be equal.

Is there a way to calculate the initial two currents that redistribute within the parallel connection, without simply dividing the input current to the parallel by the number of cells?
The issue I'm encountering is that I'm using the "Local ECM resistances" as the internal resistances, but to calculate it, I need the currents, which are my unknown variable.

**Comment by valentinsulzer** [comment]
The correct way to do it is to set up an optimization problem with "current of each cell" as your unknown variables and solve it using e.g. scipy.

Since "Local ECM resistances" does not vary *that* much as a function of current, you could also set up your own simple iterative algorithm:
1. guess uniform current
2. calculate "Local ECM resistances"
3. update currents
4. repeat, continuing until the current updates by less than your desired tolerance

The first way will be faster to solve and more robust. The second way is conceptually easier and builds better intuition for the problem

**Reply by MarcoZangrandi** [reply]
Thank you very much for your response!! 
Do you happen to know if there is already an algorithm developed in PyBaMM or Liionpack for this problem? I have implemented the second method you recommended, but I would like to make it more robust and fast, and having never worked on optimization problems, I'm not sure how to implement it. Thank you

**Comment by valentinsulzer** [comment]
If you haven't yet, you can also check out https://github.com/pybamm-team/liionpack for pack models with pybamm
